src/System.Console/src/System/Console.cs

```diff
@@ -73,6 +90,41 @@ private static TextWriter CreateOutputWriter(Stream outputStream)
                     leaveOpen: true) { AutoFlush = true });
         }
 
+        private static bool _stdInRedirectQueried = false;
+        private static bool _stdOutRedirectQueried = false;
+        private static bool _stdErrRedirectQueried = false;
+
+        private static bool _isStdInRedirected;
+        private static bool _isStdOutRedirected;
+        private static bool _isStdErrRedirected;
+
+        public static bool IsInputRedirected
+        {
+            get
+            {
+                return Volatile.Read(ref _stdInRedirectQueried) ? _isStdInRedirected
+                    : EnsureInitialized(ref _isStdInRedirected, ref _stdInRedirectQueried, () => { return ConsolePal.IsInputRedirectedCore(); });
```

Comment:
The way this is constructed, we're going to end up calling EnsureInitialized on every access to IsInputRedirected when input isn't redirected.  How about just boxing the result into a ```StrongBox<T>``` and using the existing EnsureInitialized method?  e.g.
```C#
private static StrongBox<bool> _isStdInRedirected;
...
public static bool IsInputRedirected
{
    get
    {
        StrongBox<bool> redirected = Volatile.Read(ref _isStdInRedirected) ??
            EnsureInitialized(ref _isStdInRedirected, () => new StrongBox<bool>(ConsolePal.IsInputRedirectedCore()));
        return redirected.Value;
    }
}
```